velox/common/caching/SsdFile.h

```diff
@@ -0,0 +1,297 @@
+  static int32_t regionIndex(uint64_t offset) {
+    return offset / kRegionSize;
+  }
+
+  // Updates the read count of a region.
+  void regionUsed(int32_t region, int32_t size) {
+    regionScore_[region] += size;
+  }
+
+  int32_t maxRegions() const {
+    return maxRegions_;
+  }
+
+  int32_t shardId() const {
+    return shardId_;
+  }
+
+  // Adds 'stats_' to 'stats'.
+  void updateStats(SsdCacheStats& stats) const;
+
+  // Resets this' to a post-construction empty state. See SsdCache::clear().
+  void clear();
+
+ private:
+  static constexpr int32_t kDecayInterval = 1000;
+
+  // Increments the pin count of the region of 'offset'. Caller must hold
+  // 'mutex_'.
+  void pinRegionLocked(uint64_t offset) {
+    ++regionPins_[regionIndex(offset)];
+  }
+
+  // Returns [start, size] of contiguous space for storing data of a
+  // number of contiguous 'pins' starting with the pin at index
+  // 'begin'.  Returns a run of 0 bytes if there is no space. The
+  // space does not necessarily cover all the pins, so multiple calls
+  // starting at the first unwritten pin may be needed.
+  std::pair<uint64_t, int32_t> getSpace(
+      const std::vector<CachePin>& pins,
+      int32_t begin);
+
+  // Removes all 'entries_' that reference data in regions described by
+  // 'regionIndices'.
+  void clearRegionEntriesLocked(const std::vector<int32_t>& regionIndices);
+
+  // Clears one or more  regions for accommodating new entries. The regions are
+  // added to 'writableRegions_'. Returns true if regions could be cleared.
+  bool growOrEvictLocked();
+
+  // Increments event count and periodically decays scores.
+  void newEventLocked();
+
+  // Verifies that 'entry' has the data at 'run'.
+  void verifyWrite(AsyncDataCacheEntry& entry, SsdRun run);
+
+  // Serializes access to all private data members.
+  std::mutex mutex_;
+
+  // Shard index within 'cache_'.
+  int32_t shardId_;
+
+  // Number of kRegionSize regions in the file.
+  int32_t numRegions_{0};
+
+  // True if stopped serving traffic. Happens if no evictions are
+  // possible due to everything being pinned. Clears when pins
+  // decrease and space can be cleared.
+  bool suspended_{false};
+
+  // Maximum size of the backing file in kRegionSize units.
+  const int32_t maxRegions_;
+
+  // Number of used bytes in in each region. A new entry must fit
+  // between the offset and the end of the region. This is subscripted
+  // with the region index. The regionIndex times kRegionSize is an
+  // offset into the file.
+  std::vector<uint32_t> regionSize_;
+
+  // Indices of regions available for writing new entries.
+  std::vector<int32_t> writableRegions_;
+
+  // Count of bytes read from the corresponding region. Decays with time.
+  std::vector<uint64_t> regionScore_;
+
+  // Pin count for each region.
+  std::vector<int32_t> regionPins_;
+
+  // Map of file number and offset to location in file.
+  folly::F14FastMap<SsdKey, SsdRun> entries_;
```

Comment:
All entries here are from the same file, right? Therefore, this map can be keyed on just the offset in the file.